### ISSUE TITLE
About SGD optimizer

### ISSUE THREAD

**Issue by lknight8631** [jeromerony/fast_adversarial#3]
Hi, I saw the pseudo code on the paper, but I didn't see the SGD in the algorithm.
In your implementation, I found that you use the SGD optimizer, but I didn't sure about its meaning.
`optimizer = optim.SGD([delta], lr=1)`
I'm confusing about it, thanks for your answer!!!

**Comment by jeromerony** [comment]
The PyTorch optimizer is just a way to register the variable you are optimizing and manage the learning rate with a scheduler in a lighter fashion than doing it "by hand". Also, we noticed that it is a bit faster to use an optimizer.
We use the SGD class but you may notice that there is no momentum nor weight decay, which makes it equivalent to a simple gradient descent. The main difference with the algorithm written in the paper is that the multiplier is 1 for targeted attacks instead of -1 since optimizer.step() takes a step in the direction of the negative gradient (=gradient descent) but in our case, we are doing a gradient ascent.

**Comment by lknight8631** [comment]
Thanks for your answer!!!
But after the optimizer.step(), the delta value will be changed,  you will reassign the delta.data in the last of the loops. 
Is this change will be allowed in optimization processing?

**Comment by jeromerony** [comment]
Yes: if you do the change on delta without calling data, the change will not be in-place. If the change is in-place on the underlying tensor, it does not affect the optimizer because the optimizer does not "see it". Since the updates in each iteration are independant (no momentum), it works nicely.
